core/src/main/java/org/infinispan/distribution/TriangleOrderManager.java

```diff
@@ -0,0 +1,187 @@
+ * <p>
+ * If the command topology id does not match, it throws an {@link OutdatedTopologyException}.
+ * <p>
+ * The sequence order starts with 1 and it is per segment based. This allows segments to be updated concurrently.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+public class TriangleOrderManager {
+
+   private final Map<Integer, SenderSequencer> senderSegmentMap;
+   private final Map<Integer, ReceiverSequencer> receiverSegmentMap;
+   private volatile CacheTopology currentCacheTopology;
+
+   public TriangleOrderManager() {
+      senderSegmentMap = new ConcurrentHashMap<>();
+      receiverSegmentMap = new ConcurrentHashMap<>();
+   }
+
+   public int calculateSegmentId(Object key) {
+      return currentCacheTopology.getWriteConsistentHash().getSegment(key);
+   }
+
+   public long next(int segmentId, final int commandTopologyId) {
+      checkTopologyId(commandTopologyId);
+      try {
+         return getNext(segmentId, commandTopologyId);
+      } finally {
+         //check if topology didn't change in the meanwhile
+         checkTopologyId(commandTopologyId);
+      }
+   }
+
+   public void next(Map<Integer, Long> segmentsIds, final int commandTopologyId) {
+      checkTopologyId(commandTopologyId);
+      try {
+         for (Map.Entry<Integer, Long> entry : segmentsIds.entrySet()) {
+            entry.setValue(getNext(entry.getKey(), commandTopologyId));
+         }
+      } finally {
+         //check if topology didn't change in the meanwhile
+         checkTopologyId(commandTopologyId);
+      }
+   }
+
+   public boolean isNext(int segmentId, long sequenceNumber, int commandTopologyId) {
+      final int topologyId = currentCacheTopology.getTopologyId();
+      return commandTopologyId < topologyId ||
+            (commandTopologyId == topologyId && checkIfNext(segmentId, commandTopologyId, sequenceNumber));
+   }
+
+   public boolean isNext(Map<Integer, Long> sequenceNumbers, int commandTopologyId) {
+      final int topologyId = currentCacheTopology.getTopologyId();
+      return commandTopologyId < topologyId ||
+            (commandTopologyId == topologyId && checkAllEntries(sequenceNumbers, commandTopologyId));
+   }
+
+   public void markDelivered(int segmentId, long sequenceNumber, int commandTopologyId) {
+      ReceiverSequencer sequencer = receiverSegmentMap.get(segmentId);
+      if (sequencer != null) {
+         sequencer.deliver(commandTopologyId, sequenceNumber);
+      }
+   }
+
+   public void updateCacheTopology(CacheTopology newCacheTopology) {
+      this.currentCacheTopology = newCacheTopology;
+   }
+
+   public void removeOldSegments(Collection<Integer> ownedSegments) {
+      if (ownedSegments.isEmpty()) {
+         //no segments?
+         receiverSegmentMap.clear();
+         senderSegmentMap.clear();
+      } else {
+         receiverSegmentMap.keySet().retainAll(ownedSegments);
+         senderSegmentMap.keySet().retainAll(ownedSegments);
```

Comment:
A new command with that segment will recreate the `ReceiverSequencer` anyway, even if it has a lower topology id, so I'm not sure there's any point in removing segments not owned.

Same for `SenderSequencer`, I'd rather just increment the topology id and reset the sequence number.